core/src/main/scala/kafka/log/Log.scala

```diff
@@ -83,7 +84,8 @@ case class LogAppendInfo(var firstOffset: Long,
                          targetCodec: CompressionCodec,
                          shallowCount: Int,
                          validBytes: Int,
-                         offsetsMonotonic: Boolean)
+                         offsetsMonotonic: Boolean,
+                         hasAccurateFirstOffset: Boolean)
```

Comment:
I think this is reasonable, but it feels a bit odd to have one parameter indicating whether or not we can trust another parameter, right? The fact that we abuse `firstOffset` in the first place is a big source of confusion in the code, so I'm wondering if it would be better to replace it with an `Option` to clearly express the fact that we may or may not have it. Then in places where we need to use an offset, we can write `firstOffset.getOrElse(lastOffset)`, which is more explicit and less likely to cause confusion. 

It may even be useful to have separate objects, `LeaderLogAppendInfo` which is guaranteed to have the first offset, and `ReplicaLogAppendInfo`, which may or may not have a first offset.